pennylane/decomposition/decomposition_graph.py

```diff
@@ -176,66 +245,121 @@ def __init__(
         self._start = self._graph.add_node(None)
         self._construct_graph(operations)
 
-    def _construct_graph(self, operations):
+    def _construct_graph(self, operations: Iterable[Operator | CompressedResourceOp]):
         """Constructs the decomposition graph."""
         for op in operations:
             if isinstance(op, Operator):
                 op = resource_rep(type(op), **op.resource_params)
-            idx = self._add_op_node(op)
+            idx = self._add_op_node(op, 0)
             self._original_ops_indices.add(idx)
 
-    def _add_op_node(self, op_node: CompressedResourceOp) -> int:
+    def _add_op_node(self, op: CompressedResourceOp, num_used_work_wires: int) -> int:
         """Recursively adds an operation node to the graph.
 
         An operator node is uniquely defined by its operator type and resource parameters, which
         are conveniently wrapped in a ``CompressedResourceOp``.
 
+        Args:
+            op: The operator to add to the graph.
+            num_used_work_wires: The number of wires taken from the overall budget.
+
         """
 
+        # If an operator has already been added to the graph, we return the existing node
+        # instead of creating a new one. Now when we see an operator with a different work
+        # wire budget from the one already in the graph, whether we need to create a new
+        # node for this operator is determined by whether this operator's decomposition is
+        # work-wire dependent. We have overriden __hash__ and __eq__ of the node class so
+        # that when we have a work-wire-independent operator with a different work wire
+        # budget from the existing one in the graph, the difference is ignored.
+        work_wire_dependent = op in self._work_wire_dependent_ops
```

Comment:
Maybe we could rename this `known_work_wire_dependant`? Since it could be work wire dependant, but we just don't know yet.

Reply:
Yeah I see where the confusion is, I'll try to make this section more readable.